server/src/main/java/org/elasticsearch/common/lucene/Lucene.java

```diff
@@ -869,14 +875,29 @@ public CacheHelper getReaderCacheHelper() {
                 return null; // Modifying liveDocs
             }
         }
+
         DirectoryReaderWithAllLiveDocs(DirectoryReader in) throws IOException {
-            super(in, new FilterDirectoryReader.SubReaderWrapper() {
+            super(in, new SubReaderWrapper() {
                 @Override
                 public LeafReader wrap(LeafReader leaf) {
-                    return new SubReaderWithAllLiveDocs(leaf);
+                    try {
+                        SegmentReader segmentReader = segmentReader(leaf);
+                        SegmentCommitInfo si = segmentReader.getSegmentInfo();
+                        // TODO: Replace this by hard live bits in LUCENE-8425 and add tests for rolled back
+                        if (si.getDelCount() == 0) {
+                            return new LeafReaderWithLiveDocs(segmentReader, null, segmentReader.maxDoc());
+                        } else {
+                            Bits hardLiveBits = si.info.getCodec().liveDocsFormat().readLiveDocs(si.info.dir, si, IOContext.READ);
```

Comment:
I tried to think this through and we might be subject to concurrent deletes if we do it this way. Can't we instead do something like this: 
```Java
DocIdSetIterator soft_deletes = DocValuesFieldExistsQuery.getDocValuesDocIdSetIterator("soft_deletes", sr);
          Bits liveDocs = sr.getLiveDocs();
          FixedBitSet hardLiveDocs = new FixedBitSet(numDeletes);
          hardLiveDocs.set(0, numDeletes);
          for (int i = 0; i < liveDocs.length(); i++) {
            if (liveDocs.get(i) == false) {
              if (soft_deletes.docID() < i) {
                int doc = soft_deletes.docID() == DocIdSetIterator.NO_MORE_DOCS ? DocIdSetIterator.NO_MORE_DOCS : soft_deletes.advance(i);
                if (doc != i) {
                  hardLiveDocs.clear(i);
                }
              }
            }
          }
```

note I didn't try this out.. just to provide an idea